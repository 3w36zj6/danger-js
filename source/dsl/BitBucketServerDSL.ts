// This is `danger.bitbucket_server` inside the JSON

export interface BitBucketServerJSONDSL {
  /** The related JIRA issues */
  issues: JIRAIssue[]
  /** The PR metadata */
  pr: BitBucketServerPRDSL
  /** The commits associated with the pull request */
  commits: BitBucketServerCommit[]
  /** The comments on the pull request */
  comments: BitBucketServerPRActivity[]
  /** The activities such as OPENING, CLOSING, MERGING or UPDATING a pull request */
  activities: BitBucketServerPRActivity[]
}

// This is `danger.bitbucket_server`

/** The BitBucketServer metadata for your PR */
export interface BitBucketServerDSL extends BitBucketServerJSONDSL {}

/**
 * This is `danger.bitbucket_server.issues` It refers to the issues that are linked to the Pull Request.
 */
export interface JIRAIssue {
  key: string
  url: string
}

// This is `danger.bitbucket_server.pr`

/**
 * An exact copy of the PR's reference JSON. This interface has type'd the majority
 * of it for tooling's sake, but any extra metadata which BitBucket Server send
 * will still be inside the JS object.
 */

export interface BitBucketServerPRDSL {
  id: number
  version: number
  title: string
  description: string
  state: "OPEN" | "MERGED" | "DECLINED"
  open: boolean
  closed: boolean
  createdDate: number
  updatedDate: number
  fromRef: BitBucketServerMergeRef
  toRef: BitBucketServerMergeRef
  locked: boolean
  author: BitBucketServerPRParticipant & { role: "AUTHOR" }
  reviewers: (BitBucketServerPRParticipant & { role: "REVIEWER" })[]
  participants: (BitBucketServerPRParticipant & { role: "PARTICPANT" })[]
  links: BitBucketServerLinks<"self">
}

// These are the individual subtypes of objects inside the larger DSL objects above.

/** A BitBucketServer specific implmentation of a git commit. */
export interface BitBucketServerCommit {
  id: string
  displayId: string
  author: {
    name: string
    emailAddress: string
  }
  authorTimestamp: number
  committer: {
    name: string
    emailAddress: string
  }
  committerTimestamp: number
  message: string
  parents: {
    id: string
    displayId: string
  }[]
}

export type BitBucketServerPRStatus = "APPROVED" | "UNAPPROVED" | "NEEDS_WORK"

export interface BitBucketServerPRParticipant {
  user: BitBucketServerUser
  role: "AUTHOR" | "REVIEWER" | "PARTICIPANT"
  approved: boolean
  status: BitBucketServerPRStatus
}

/**
 * A BitBucketServer user account.
 */
export interface BitBucketServerUser {
  name: string
  emailAddress: string
  id: number
  displayName: string
  active: boolean
  slug: string
  type: "NORMAL"
}

/**
 * A BitBucket Server Repo
 */
export interface BitBucketServerRepo {
  slug: string
  name?: string
  scmId: string
  public: boolean
  forkable: boolean
  links: BitBucketServerLinks<"self" | "clone">

  project: {
    id: number
    key: string
    public: boolean
    name: string
    type: string
    links: BitBucketServerLinks<"self">
  }
}

export type BitBucketServerLinks<Names> = {
  [key in keyof Names]: {
    href: string
    name?: string
  }[]
}

export interface BitBucketServerMergeRef {
  id: string
  displayId: string
  latestCommit: string
  repository: BitBucketServerRepo
}

export interface BitBucketServerPRActivity {
  action: "COMMENTED" | "OPENED" | "MERGED" | "DECLINED" | "UPDATED"
  comment?: BitBucketServerPRComment
  commentAction?: "ADDED" | "UPDATED"
  commentAnchor?: {
    diffType: "COMMIT" | "EFFECTIVE" | "REQUIRED" | "RANGE"
    line: number
    lineType: "CONTEXT" | "ADDED" | "REMOVED"
    fileType: "FROM" | "TO"
    fromHash: string
    path: string
    srcPath: string
    toHash: string
  }
  createdDate: number
  id: number
  user: BitBucketServerUser
}

export interface BitBucketServerPRComment {
  author: BitBucketServerUser
  comments: BitBucketServerPRActivity[]
  createdDate: number
  updatedDate: number
  id: number
  permittedOperations: {
    deletable: boolean
    editable: boolean
  }
  text: string
  version: number
  parent?: {
    id: number
  }
}
