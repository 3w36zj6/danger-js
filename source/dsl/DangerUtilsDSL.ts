/**
 * The Danger Utils DSL contains utility functions
 * that are specific to universal Danger use-cases.
 */
export interface DangerUtilsDSL {

  /**
   * Creates an HTML link.
   *
   * @param {string} href The HTML link's destination
   * @param {string} text The HTML link's text
   * @returns {string} The HTML <a> tag
   */
  href(href: string, text: string): string

  /**
   * Converts an array of strings into a sentence.
   *
   * @param {Array<string>} array The array of strings
   * @returns {string} The sentence
   */
  sentence(array: Array<string>): string
}
