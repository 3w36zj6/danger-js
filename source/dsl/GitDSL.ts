import { GitCommit } from "./Commit"

/** The results of running a JSON patch */
export interface JSONPatch {
  /** The JSON in a file at the PR merge base */
  before: any,
  /** The JSON in a file from the PR submitter */
  after: any,
  /** The set of operations to go from one JSON to another JSON */
  diff: Array<JSONPatchOperation>
}

/** An individual operation inside an rfc6902 JSON Patch */
export interface JSONPatchOperation {
  /** An operation type */
  op: string
  /** The JSON keypath which the operation applies on */
  path: string
  /** The changes for applied */
  value: string
}

/** All JSON diff values will be this shape */
export interface JSONDiffValue {
  /** The value before the PR's applied changes */
  before: any
  /** The value after the PR's applied changes */
  after: any
  /** If both before & after are arrays, then you optionally get what is added. Emprty is no additional objects. */
  added?: any[],
  /** If both before & after are arrays, then you optionally get what is removed. Emprty is no removed objects. */
  removed?: any[]
}

// This is `danger.git`

/** The git specific metadata for a PR */
export interface GitDSL {
  /**
   * Filepaths with changes relative to the git root
   * @type {string[]}
   */
  readonly modified_files: Readonly<Array<string>>

  /**
   * Newly created filepaths relative to the git root
   * @type {string[]}
   */
  readonly created_files: Readonly<Array<string>>

  /**
   * Removed filepaths relative to the git root
   * @type {string[]}
   */
  readonly deleted_files: Readonly<Array<string>>

  /** Offers the diff for a specific file */
  diffForFile(filename: string): string | null,

  /** Provides a JSON patch (rfc6902) between the two versions of a JSON file, returns null if you don't have any changes for the file in the diff. */
  JSONPatchForFile(filename: string): Promise<JSONPatch | null>,

  /** Provides a simplified JSON diff between the two versions of a JSON file */
  JSONDiffForFile(filename: string): Promise<Array<any>>,

  /** The Git commit metadata */
  readonly commits: Readonly<Array<GitCommit>>
}
