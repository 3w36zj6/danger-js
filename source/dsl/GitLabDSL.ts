// Please don't have includes in here that aren't inside the DSL folder, or the d.ts/flow defs break
import { Gitlab, Types } from "@gitbeaker/node"
import { RepoMetaData } from "./RepoMetaData"

// getPlatformReviewDSLRepresentation
export interface GitLabJSONDSL {
  /** Info about the repo */
  metadata: RepoMetaData
  /** Info about the merge request */
  mr: GitLabMR
  /** All the individual commits in the merge request */
  commits: Types.CommitSchema[]
  /** Merge Request-level MR approvals Configuration */
  approvals: Types.MergeRequestLevelMergeRequestApprovalSchema
}

// danger.gitlab
/** The GitLab metadata for your MR */
export interface GitLabDSL extends GitLabJSONDSL {
  utils: {
    fileContents(path: string, repoSlug?: string, ref?: string): Promise<string>
    addLabels(...labels: string[]): Promise<boolean>
    removeLabels(...labels: string[]): Promise<boolean>
  }
  api: InstanceType<typeof Gitlab>
}

// ---
// JSON responses from API
export interface GitlabUpdateMr extends Types.UpdateMergeRequestOptions, Types.BaseRequestOptions {}

export interface GitLabNote extends Types.MergeRequestNoteSchema {
  type: "DiffNote" | "DiscussionNote" | null // XXX: other types? null means "normal comment"
}

export interface GitLabInlineNote extends GitLabNote {
  type: "DiffNote" | "DiscussionNote" | null // XXX: other types? null means "normal comment"
}

/** TODO: These need more comments from someone who uses GitLab, see GitLabDSL.ts in the danger-js repo */
export interface GitLabMR extends Types.MergeRequestSchema {
  subscribed: boolean
  changes_count: string
  latest_build_started_at: string
  latest_build_finished_at: string
  first_deployed_to_production_at: string | null
  pipeline: {
    id: number
    sha: string
    ref: string
    status: "canceled" | "failed" | "pending" | "running" | "skipped" | "success"
    web_url: string
  }
  diff_refs: {
    base_sha: string
    head_sha: string
    start_sha: string
  }
  diverged_commits_count: number
  rebase_in_progress: boolean
  approvals_before_merge: null
  //
  /** Access rights for the user who created the MR */
  user: {
    /** Does the author have access to merge? */
    can_merge: boolean
  }
  merge_error: null
  allow_collaboration: boolean
  allow_maintainer_to_push: boolean
}

export interface GitLabDiscussion {
  id: string //40 character hex
  individual_note: boolean
  notes: GitLabNote[]
}

export interface GitLabDiscussionTextPosition {
  position_type: "text"
  base_sha: string
  start_sha: string
  head_sha: string
  new_path: string
  new_line: number
  old_path: string
  old_line: number | null
}

export interface GitLabDiscussionCreationOptions {
  position?: GitLabDiscussionTextPosition
}

export interface GitLabInlineNote extends GitLabNote {
  position: {
    base_sha: string
    start_sha: string
    head_sha: string
    old_path: string
    new_path: string
    position_type: "text" // XXX: other types?
    old_line: number | null
    new_line: number
  }
  resolvable: boolean
  resolved: boolean
  resolved_by: GitLabUser | null
}

export interface GitLabMRCommit {
  id: string
  short_id: string
  created_at: string
  parent_ids: string[]
  title: string
  message: string
  author_name: string
  author_email: string
  authored_date: string
  committer_name: string
  committer_email: string
  committed_date: string
}

export interface GitLabCommit {
  id: string
  short_id: string
  title: string
  author_name: string
  author_email: string
  created_at: string
}

export interface GitLabRepositoryCompare {
  commit: GitLabCommit
  commits: GitLabCommit[]
  diffs: GitLabMRChange[]
  compare_timeout: boolean
  compare_same_ref: boolean
}

export interface GitLabApproval {
  id: number
  iid: number
  project_id: number
  title: string
  description: string
  state: "closed" | "open" | "locked" | "merged"
  created_at: string
  updated_at: string
  merge_status: "can_be_merged"
  approvals_required: number
  approvals_left: number
  approved_by?:
    | {
        user: GitLabUser
      }[]
    | GitLabUser[]
}
