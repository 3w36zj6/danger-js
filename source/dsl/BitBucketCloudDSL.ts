import { BitBucketServerPRDSL, RepoMetaData as BitBucketServerRepoMetaData } from "./BitBucketServerDSL"

export interface BitBucketCloudPagedResponse<T> {
  pagelen: number
  size: number
  page: number
  next: string | undefined
  previous: string | undefined
  values: T[]
}

export type BitBucketCloudPRDSL = BitBucketServerPRDSL
export type RepoMetaData = BitBucketServerRepoMetaData

export interface BitBucketCloudUser {
  /** The uuid of the commit author */
  uuid: string

  /** The display name of the commit author */
  display_name: string

  /** The nick name of the commit author */
  nickname: string

  /** The acount id of the commit author */
  account_id: string
}

/** A BitBucketServer specific implementation of a git commit. */
export interface BitBucketCloudCommit {
  /** The SHA for the commit */
  hash: string

  /** The author of the commit, assumed to be the person who wrote the code. */
  author: BitBucketCloudUser

  /** When the commit was commited to the project, in ISO 8601 format */
  date: string
  /** The commit's message */
  message: string
  /** The commit's parents */
  parents: {
    /** The full SHA */
    hash: string
  }[]
}

export interface BitBucketCloudPRLink {
  id: number
  links: {
    self: {
      href: string
    }
    html: {
      href: string
    }
  }
  title: string
}

export interface BitBucketCloudContent {
  raw: string
  markup: string
  html: string
  type: "rendered"
}

export interface BitBucketCloudPRComment {
  deleted: boolean
  pullrequest: BitBucketCloudPRLink
  content: BitBucketCloudContent

  /** When the comment was created, in ISO 8601 format */
  created_on: string
  user: BitBucketCloudUser

  /** When the comment was updated, in ISO 8601 format */
  updated_on: string
  type: string
  id: number
}

export interface BitBucketCloudPRActivity {
  comment?: BitBucketCloudPRComment
  pull_request: {
    id: number
    title: string
  }
}
