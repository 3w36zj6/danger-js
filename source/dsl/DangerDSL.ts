import { GitDSL, GitJSONDSL } from "../dsl/GitDSL"
import { GitHubDSL } from "../dsl/GitHubDSL"
import { DangerUtilsDSL } from "./DangerUtilsDSL"

/**
 *  The root of the Danger JSON DSL.
 */

export interface DangerDSLJSONType {
  /** The data only version of Git DSL */
  git: GitJSONDSL
  /** The data only version of GitHub DSL */
  github: GitHubDSL
}

/**
 *  The Danger DSL provides the metadata for introspection
 *  in order to create your own rules.
 */
export interface DangerDSLType {
  /**
   *  Details specific to the git changes within the code changes.
   *  Currently, this is just the raw file paths that have been
   *  added, removed or modified.
   */
  readonly git: GitDSL
  /**
   *  The GitHub metadata. This covers things like PR info,
   *  comments and reviews on the PR, label metadata, commits with
   *  GitHub user identities and some useful utility functions
   *  for displaying links to files.
   *
   *  Also provides an authenticated API so you can work directly
   *  with the GitHub API. That is an instance of the "github" npm module.
   */
  readonly github: GitHubDSL

  /**
   * Functions which are gloablly useful in most Dangerfiles. Right
   * now, these functions are around making sentences of arrays, or
   * for making hrefs easily.
   */
  readonly utils: DangerUtilsDSL
}

/// End of Danger DSL definition

export class DangerDSL {
  public readonly github: GitHubDSL

  constructor(platformDSL: any, public readonly git: GitDSL, public readonly utils: DangerUtilsDSL) {
    // As GitLab etc support is added this will need to be changed
    this.github = platformDSL
  }
}
