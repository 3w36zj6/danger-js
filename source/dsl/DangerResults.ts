import { Violation, isInline } from "../dsl/Violation"

/**
 * The representation of what running a Dangerfile generates.
 *
 * In the future I'd like this to be cross process, so please
 * do not add functions, only data to this interface.
 */
export interface DangerResults {
  /**
   * Failed messages
   */
  fails: Violation[]

  /**
   * Messages for info
   */
  warnings: Violation[]

  /**
   * A set of messages to show inline
   */
  messages: Violation[]

  /**
   * Markdown messages to attach at the bottom of the comment
   */
  markdowns: Violation[]
}

export interface DangerRuntimeContainer extends DangerResults {
  /**
   * Asynchronous functions to be run after parsing
   */
  scheduled?: any[]
}

export interface DangerInlineResults {
  /**
   * Path to the file
   *
   * @type {string}
   */
  file: string

  /**
   * Line in the file
   *
   * @type {string}
   */

  line: number

  /**
   * Failed messages
   */
  fails: string[]

  /**
   * Messages for info
   */
  warnings: string[]

  /**
   * A set of messages to show inline
   */
  messages: string[]

  /**
   * Markdown messages to attach at the bottom of the comment
   */
  markdowns: string[]
}

export const emptyDangerResults = {
  fails: [],
  warnings: [],
  messages: [],
  markdowns: [],
}

export function inlineResults(results: DangerResults): DangerResults {
  return {
    fails: results.fails.filter(m => isInline(m)),
    warnings: results.warnings.filter(m => isInline(m)),
    messages: results.messages.filter(m => isInline(m)),
    markdowns: results.markdowns.filter(m => isInline(m)),
  }
}

export function regularResults(results: DangerResults): DangerResults {
  return {
    fails: results.fails.filter(m => !isInline(m)),
    warnings: results.warnings.filter(m => !isInline(m)),
    messages: results.messages.filter(m => !isInline(m)),
    markdowns: results.markdowns.filter(m => !isInline(m)),
  }
}

export function mergeResults(results1: DangerResults, results2: DangerResults): DangerResults {
  return {
    fails: results1.fails.concat(results2.fails),
    warnings: results1.warnings.concat(results2.warnings),
    messages: results1.messages.concat(results2.messages),
    markdowns: results1.markdowns.concat(results2.markdowns),
  }
}

export function sortInlineResults(inlineResults: DangerInlineResults[]): DangerInlineResults[] {
  // First sort messages in every inline result
  const sortedInlineResults = inlineResults.map(i => {
    return {
      file: i.file,
      line: i.line,
      fails: i.fails.sort(),
      warnings: i.warnings.sort(),
      messages: i.messages.sort(),
      markdowns: i.markdowns.sort(),
    }
  })

  // Then sort a whole array of inline results based on file/line
  return sortedInlineResults.sort((a, b) => {
    if (a.file < b.file) {
      return -1
    } else if (a.file > b.file) {
      return 1
    } else if (a.line < b.line) {
      return -1
    } else if (a.line > b.line) {
      return 1
    } else {
      // both file & line are the same
      return 0
    }
  })
}

export function sortResults(results: DangerResults): DangerResults {
  const sortByFile = (a: Violation, b: Violation): number => {
    if (a.file === undefined) {
      return -1
    }
    if (b.file === undefined) {
      return 1
    }

    if (a.file == b.file) {
      if (a.line == undefined) {
        return -1
      }
      if (b.line == undefined) {
        return 1
      }

      if (a.line < b.line) {
        return -1
      } else if (a.line > b.line) {
        return 1
      } else {
        return 0
      }
    }

    if (a.file < b.file) {
      return -1
    } else {
      return 1
    }
  }

  return {
    fails: results.fails.sort(sortByFile),
    warnings: results.warnings.sort(sortByFile),
    messages: results.messages.sort(sortByFile),
    markdowns: results.markdowns.sort(sortByFile),
  }
}

export function resultsIntoInlineResults(results: DangerResults): DangerInlineResults[] {
  // Here we iterate through all keys ("fails", "warnings", "messages", "markdowns") and for each violation
  // in given kind we produce new DangerInlineResult or append a violation to existing result. This is all
  // happening in a `violationsIntoInlineResults` function that mutates an out-of-scope variable `dangerInlineResults`.

  const dangerInlineResults: DangerInlineResults[] = []

  const violationsIntoInlineResults = (kind: string) => {
    for (let violation of results[kind]) {
      if (violation.file && violation.line) {
        const findInlineResult = dangerInlineResults.find(r => r.file == violation.file && r.line == violation.line)
        if (findInlineResult) {
          findInlineResult[kind].push(violation.message)
        } else {
          const inlineResult = {
            file: violation.file,
            line: violation.line,
            fails: [],
            warnings: [],
            messages: [],
            markdowns: [],
          }
          inlineResult[kind].push(violation.message)
          dangerInlineResults.push(inlineResult)
        }
      }
    }
  }
  Object.keys(results).forEach(violationsIntoInlineResults)

  return dangerInlineResults
}

export function inlineResultsIntoResults(inlineResults: DangerInlineResults): DangerResults {
  const messageToViolation = (message: string): Violation => {
    return { message: message, file: inlineResults.file, line: inlineResults.line }
  }

  return {
    fails: inlineResults.fails.map(messageToViolation),
    warnings: inlineResults.warnings.map(messageToViolation),
    messages: inlineResults.messages.map(messageToViolation),
    markdowns: inlineResults.markdowns.map(messageToViolation),
  }
}
