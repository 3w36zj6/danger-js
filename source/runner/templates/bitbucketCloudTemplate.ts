import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"

// This unicode emojis also work for old versions of bitbucket server, were emojis are not supported
const noEntryEmoji = "❌"
const warningEmoji = "⚠️"
const messageEmoji = "✨"

export const dangerSignature = (results: DangerResults) => {
  let meta = results.meta || { runtimeName: "dangerJS", runtimeHref: "https://danger.systems/js" }
  return `Generated by ${warningEmoji} [${meta.runtimeName}](${meta.runtimeHref})`
}

export const messageForResultWithIssues = `${warningEmoji} Danger found some issues. Don't worry, everything is fixable.`

export const dangerIDToString = (id: string) => `danger-id-${id};`

/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = (results: DangerResults, commitID: string) => `


  > _${dangerSignature(results)} against ${commitID}
  `

function buildMarkdownTable(header: string, emoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return `

  |      ${violations.length} ${header} |
  | --- |
  ${violations.map(v => `| ${emoji} ${v.message} |`)}

  `
}

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {string} commitID The hash that represents the latest commit
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(dangerID: string, commitID: string, results: DangerResults): string {
  return `
  ${buildMarkdownTable("Fails", noEntryEmoji, results.fails)}
  ${buildMarkdownTable("Warnings", warningEmoji, results.warnings)}
  ${buildMarkdownTable("Messages", messageEmoji, results.messages)}
  
  ${results.markdowns.map(v => v.message).join("\n\n")}
  
  ${dangerSignaturePostfix(results, commitID)}
  
  [](http://${dangerIDToString(dangerID)})
  `
}
