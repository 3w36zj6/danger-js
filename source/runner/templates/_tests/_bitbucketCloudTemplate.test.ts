import {
  emptyResults,
  failsResultsWithoutMessages,
  warnResults,
  failsResults,
  messagesResults,
  markdownResults,
  summaryResults,
  multipleSummaryResults,
} from "../../_tests/fixtures/ExampleDangerResults"
import { dangerSignaturePostfix, template, inlineTemplate } from "../bitbucketCloudTemplate"
import { DangerResults } from "../../../dsl/DangerResults"

const commitID = "e70f3d6468f61a4bef68c9e6eaba9166b096e23c"

const noEntryEmoji = "❌"
const warningEmoji = "⚠️"
const messageEmoji = "✨"

describe("generating messages for BitBucket cloud", () => {
  it("shows no sections for empty results", () => {
    const issues = template("blankID", commitID, emptyResults)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("shows no sections for results without messages", () => {
    const issues = template("blankID", commitID, failsResultsWithoutMessages)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("Shows the failing messages in a section", () => {
    const issues = template("blankID", commitID, failsResults)
    expect(issues).toContain("Fails")
    expect(issues).not.toContain("Warnings")
  })

  it("Shows the warning messages in a section", () => {
    const issues = template("blankID", commitID, warnResults)
    expect(issues).toContain("Warnings")
    expect(issues).not.toContain("Fails")
  })

  it("summary result matches snapshot", () => {
    expect(template("blankID", commitID, summaryResults)).toMatchSnapshot()
  })

  it("multiple summary result matches snapshot", () => {
    expect(template("blankID", commitID, multipleSummaryResults)).toMatchSnapshot()
  })

  it("shows a postfix message indicating the current commit ID at the time of comment", () => {
    expect(template("blankID", commitID, emptyResults)).toContain(dangerSignaturePostfix({} as DangerResults, commitID))
  })
})

describe("generating inline messages", () => {
  it("Shows the failing message", () => {
    const issues = inlineTemplate("blankID", failsResults, "File.swift", 5)
    expect(issues).toContain(`- ${noEntryEmoji} Failing message`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
    expect(issues).not.toContain(`- ${messageEmoji}`)
  })

  it("Shows the warning message", () => {
    const issues = inlineTemplate("blankID", warnResults, "File.swift", 5)
    expect(issues).toContain(`- ${warningEmoji} Warning message`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${messageEmoji}`)
  })

  it("Shows the message", () => {
    const issues = inlineTemplate("blankID", messagesResults, "File.swift", 5)
    expect(issues).toContain(`- ${messageEmoji} Message`)
    expect(issues).not.toContain(`- ${noEntryEmoji}`)
    expect(issues).not.toContain(`- ${warningEmoji}`)
  })

  it("Shows markdowns one after another", () => {
    const issues = inlineTemplate("blankID", markdownResults, "File.swift", 5)
    const expected = `
### Short Markdown Message1

### Short Markdown Message2
`
    expect(issues).toContain(expected)
  })

  it("summary inline result matches snapshot", () => {
    expect(inlineTemplate("blankID", summaryResults, "File.swift", 5)).toMatchSnapshot()
  })
})
