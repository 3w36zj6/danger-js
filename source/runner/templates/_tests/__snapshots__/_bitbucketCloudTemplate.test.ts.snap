// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating messages for BitBucket cloud summary result matches snapshot 1`] = `
"
  ⚠️  Danger found some issues. Don't worry, everything is fixable.

  

  |      1 Fails |
  | --- |
  | ❌ Failing message Failing message |

  
  

  |      1 Warnings |
  | --- |
  | ⚠️ Warning message Warning message |

  
  

  |      1 Messages |
  | --- |
  | ✨ message |

  
  
  markdown
  
  


  _Generated by ⚠️ [dangerJS](https://danger.systems/js) against e70f3d6468f61a4bef68c9e6eaba9166b096e23c
  
  
  [](http://danger-id-blankID;)
  "
`;
